### PR TITLE
[MC-1518] Rank using region engagement in experiment

### PR DESCRIPTION
**Pull request by mmiermans**
## References

JIRA: [MC-1518](https://mozilla-hub.atlassian.net/browse/MC-1518)
[Experiment brief](https://docs.google.com/document/d/1YaZYPZ-ZR06_9XdNcm27KUDSvGi9hl_51G0vp6ol-Uk/edit) (in progress of being drafted)

## Description
Building on https://github.com/mozilla-services/merino-py/pull/633, this PR uses regional engagement in Thompson sampling in the treatment branch of an experiment.

## PR Review Checklist

_Put an `x` in the boxes that apply_

- [x] This PR conforms to the [Contribution Guidelines](https://github.com/mozilla-services/merino-py/blob/main/CONTRIBUTING.md)
- [x] The PR title starts with the JIRA issue reference, format `[DISCO-####]`, and has the same title (if applicable)
- [ ] `[load test: (abort|warn)]` keywords are applied (if applicable)
- [] [Documentation](https://github.com/mozilla-services/merino-py/tree/main/docs) has been updated (if applicable)
- [x] [Functional and performance test](https://github.com/mozilla-services/merino-py/blob/main/docs/dev/testing.md) coverage has been expanded and maintained (if applicable)

[MC-1518]: https://mozilla-hub.atlassian.net/browse/MC-1518?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: merino/curated_recommendations/provider.py [hunk 1]
```diff
@@ -149,7 +157,7 @@ def process_recommendations(
 
     async def fetch(
         self, curated_recommendations_request: CuratedRecommendationsRequest
-    ) -> CuratedRecommendationsResponse:  # noqa
+    ) -> CuratedRecommendationsResponse:
```
Comment (mmiermans) [inline merino/curated_recommendations/provider.py:160]: I don't know why there was a `noqa` here.

File: merino/curated_recommendations/rankers.py [hunk 2]
```diff
@@ -11,6 +11,13 @@
 DEFAULT_BETA_PRIOR = (
     12500  # 0.5% of median German NewTab item impressions for 2023-03-28 to 2023-04-05.
 )
```
Comment (mmiermans) [inline merino/curated_recommendations/rankers.py:13]: There's a separate task [MC-1460](https://mozilla-hub.atlassian.net/browse/MC-1460) to update the priors. That's out-of-scope for this PR.

File: merino/curated_recommendations/protocol.py [hunk 3]
```diff
@@ -94,6 +104,11 @@ class CuratedRecommendationsRequest(BaseModel):
     count: int = 100
     topics: list[Topic | str] | None = None
     feeds: list[str] | None = None
+    # Firefox sends the name and branch for Nimbus experiments on the "pocketNewtab" feature:
+    # https://searchfox.org/mozilla-central/source/browser/components/newtab/lib/DiscoveryStreamFeed.sys.mjs
+    # Allow any string value or null, because ExperimentName is not meant to be an exhaustive list.
+    experimentName: ExperimentName | str | None = None
+    experimentBranch: str | None = None
```
Comment (lfishhead) [inline merino/curated_recommendations/protocol.py:111]: this is good to know.  for eventual reporting we get the nimbus info on engagement events?

File: merino/curated_recommendations/rankers.py [hunk 4]
```diff
@@ -29,24 +39,43 @@ def thompson_sampling(
...
+
+        # Use a weighted average of regional and global engagement, if that's enabled and available.
+        if enable_region_engagement:
+            region_opens, region_no_opens = get_opens_no_opens(rec, region)
+            if region_no_opens:
+                opens = (region_weight * region_opens) + ((1 - region_weight) * opens)
+                no_opens = (region_weight * region_no_opens) + ((1 - region_weight) * no_opens)
+
+        opens += alpha_prior
+        no_opens += beta_prior
```
Comment (lfishhead) [inline merino/curated_recommendations/rankers.py:73]: just comments here.  this is a more interesting approach than I had understood.  I had thought you maintained two separate models for CTR and were combining samples from each using the weightings.  This is a single model that weights (~20x) the regional engagement based on avg impression volume.  they are likely similar in practice.  Either way this looks good, tho I am hopeful we will experiment with a more scalable approach to contextual variables in the future.

also every time I see this code it gets cleaner which I appreciate

**Review by lfishhead — APPROVED** [review]

File: merino/curated_recommendations/protocol.py (hunk 3 above)
Comment (mmiermans) [reply merino/curated_recommendations/protocol.py:111]: Yes, these are the Nimbus experiment and branch name. There are a bunch of existing Nimbus metrics, for example the mean CTR, and % of users who click on any recommendation. Data Science can create new metrics.

**Pull request merged**